lib/accesslists/hierarchy.go

```diff
@@ -0,0 +1,837 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accesslists
+
+import (
+	"context"
+	"slices"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/accesslist"
+	"github.com/gravitational/teleport/api/types/trait"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// RelationshipKind represents the type of relationship: member or owner.
+type RelationshipKind int
+
+const (
+	RelationshipKindMember RelationshipKind = iota
+	RelationshipKindOwner
+)
+
+// MembershipOrOwnershipType represents the type of membership or ownership a User has for an Access List.
+type MembershipOrOwnershipType int
+
+const (
+	// MembershipOrOwnershipTypeNone indicates that the User lacks valid Membership or Ownership for the Access List.
+	MembershipOrOwnershipTypeNone MembershipOrOwnershipType = iota
+	// MembershipOrOwnershipTypeExplicit indicates that the User has explicit Membership or Ownership for the Access List.
+	MembershipOrOwnershipTypeExplicit
+	// MembershipOrOwnershipTypeInherited indicates that the User has inherited Membership or Ownership for the Access List.
+	MembershipOrOwnershipTypeInherited
+)
+
+type MembersAndLocksGetter interface {
+	ListAccessListMembers(ctx context.Context, accessListName string, pageSize int, pageToken string) (members []*accesslist.AccessListMember, nextToken string, err error)
+}
+
+// HierarchyNode represents an access list and its relationships.
+type HierarchyNode struct {
+	// AccessList is the Underlying AccessList object.
+	AccessList *accesslist.AccessList
+	// MemberUsers are users who are direct members of this list.
+	MemberUsers map[string]*accesslist.AccessListMember
+	// MemberLists are AccessLists that are direct members of this list.
+	MemberLists map[string]*HierarchyNode
+	// OwnerUsers are users who are direct Owners of this list.
+	OwnerUsers map[string]*accesslist.Owner
+	// OwnerLists are AccessLists that are direct Owners of this list.
+	OwnerLists map[string]*HierarchyNode
+	// MemberParents are AccessLists that have this list as a member.
+	MemberParents map[string]*HierarchyNode
+	// OwnerParents are AccessLists that have this list as an owner.
+	OwnerParents map[string]*HierarchyNode
+}
+
+type hierarchy struct {
+	Nodes map[string]*HierarchyNode
+	Locks services.LockGetter
+	Clock clockwork.Clock
+}
+
+type Hierarchy interface {
```

Comment:
Nit: Feels like this being the central interface for interacting with nested access lists should have a good godoc on it.

Reply:
added one, WDYT? the funcs have more detailed docs themselves, not sure what else to add to the interface doc

Reply:
Looks good to me!